### PR TITLE
ENH: Update scipy.interpolate section

### PR DESCRIPTION
**Pull request by mdhaber**
This is a fairly minimal update to the interpolate section. Important points are that the example no longer uses the legacy function `interp1d`, and it highlights the distinction between the two main uses of splines: smoothing and interpolation.

One thing I could consider doing including in this PR is to simplify variable names. I maintained the existing style, but I would find the code easier to read with simpler variable names like `t` and `x`. LMK what you think.

In follow-up PRs, I could:
- Show how to take the `derivative` and `antiderivative` of a spline in the text.
- Show how to get the roots of a spline in the text.
- Create a separate example that shows how to create an inverse of a function.

LMK what you'd be interested in.

### CODE REVIEW

File: intro/scipy/examples/plot_interpolation.py [hunk 1]
```diff
@@ -9,25 +9,48 @@
 import numpy as np
 
 rng = np.random.default_rng(27446968)
-measured_time = np.linspace(0, 1, 10)
-noise = 1e-1 * (rng.random(10) * 2 - 1)
-measures = np.sin(2 * np.pi * measured_time) + noise
+measured_time = np.linspace(0, 2*np.pi, 20)
```
Comment (mdhaber) [inline intro/scipy/examples/plot_interpolation.py:12]:
```suggestion
measured_time = np.linspace(0, 2 * np.pi, 20)
```

File: intro/scipy/examples/plot_interpolation.py [hunk 2]
```diff
@@ -9,25 +9,48 @@
...
+# Smooth the curve and interpolate at new times
 import scipy as sp
 
-linear_interp = sp.interpolate.interp1d(measured_time, measures)
-interpolation_time = np.linspace(0, 1, 50)
-linear_results = linear_interp(interpolation_time)
-cubic_interp = sp.interpolate.interp1d(measured_time, measures, kind="cubic")
-cubic_results = cubic_interp(interpolation_time)
+smoothing_spline = sp.interpolate.make_smoothing_spline(measured_time, measurements)
+interpolation_time = np.linspace(0, 2*np.pi, 200)
```
Comment (mdhaber) [inline intro/scipy/examples/plot_interpolation.py:21]:
```suggestion
interpolation_time = np.linspace(0, 2 * np.pi, 200)
```

File: intro/scipy/examples/plot_interpolation.py [hunk 3]
```diff
@@ -9,25 +9,48 @@
...
 import matplotlib.pyplot as plt
 
 plt.figure(figsize=(6, 4))
-plt.plot(measured_time, measures, "o", ms=6, label="measures")
-plt.plot(interpolation_time, linear_results, label="linear interp")
-plt.plot(interpolation_time, cubic_results, label="cubic interp")
+plt.plot(measured_time, measurements, ".", ms=6, label="measurements")
+plt.plot(interpolation_time, smooth_results, label="smoothing spline")
+plt.plot(interpolation_time, np.sin(interpolation_time),
+         '--', label="underlying curve")
```
Comment (mdhaber) [inline intro/scipy/examples/plot_interpolation.py:31]:
```suggestion
plt.plot(interpolation_time, np.sin(interpolation_time), "--", label="underlying curve")
```

File: intro/scipy/examples/plot_interpolation.py [hunk 4]
```diff
@@ -9,25 +9,48 @@
...
+# Fit the data exactly
+interp_spline = sp.interpolate.make_interp_spline(measured_time, function)
+interp_results = interp_spline(interpolation_time)
+
+# Plot the data, the interpolant, and the original function
+plt.figure(figsize=(6, 4))
+plt.plot(measured_time, function, ".", ms=6, label="measurements")
+plt.plot(interpolation_time, interp_results, label="interpolating spline")
+plt.plot(interpolation_time, np.sin(interpolation_time),
+         '--', label="underlying curve")
```
Comment (mdhaber) [inline intro/scipy/examples/plot_interpolation.py:44]:
```suggestion
plt.plot(interpolation_time, np.sin(interpolation_time), "--", label="underlying curve")
```

File: intro/scipy/examples/plot_interpolation.py [hunk 5]
```diff
@@ -9,25 +9,48 @@
...
+plt.legend()
+plt.show()
+
+# Plot interpolant, its derivative, and its antiderivative
+plt.figure(figsize=(6, 4))
+t = interpolation_time
+plt.plot(t, interp_spline(t), label="spline")
+plt.plot(t, interp_spline.derivative()(t), label="derivative")
+plt.plot(t, interp_spline.antiderivative()(t)-1, label="antiderivative")
```
Comment (mdhaber) [inline intro/scipy/examples/plot_interpolation.py:53]:
```suggestion
plt.plot(t, interp_spline.antiderivative()(t) - 1, label="antiderivative")
```

**Review by mdhaber — COMMENTED** [review]
Adjustments to appease `black`.

**Review by jarrodmillman — APPROVED** [review]

**Pull request merged**